README.md

<!-- README.md is generated from README.Rmd. Please edit that file -->

# ![](man/figures/header.png)

<!-- badges: start -->

![](http://img.shields.io/badge/cool-useless-green.svg)
<!-- badges: end -->

The `exhibitionist` package provides functions for showing and
annotating sequences of characters or numbers.

## What’s in the box

  - `show_bit()` for showing bit sequences of integers and doubles in
    the console
  - `plot_char()` for plotting a sequence of characters as defined in a
    `chars_df` data.frame with an `annotation_df` data.frame used to
    control the annotation underneat the characters.

## Installation

You can install from
[GitHub](https://github.com/coolbutuseless/exhibitionist) with:

``` r
# install.packages("devtools")
devtools::install_github("coolbutuseless/exhibitionist")
```

# `show_bit()` for displaying integers and doubles in the terminal

``` r
show_bits(12345.67)
#> 0 10000001100 (13) 1000000111001101010111000010100011110101110000101001 : 12345.67
```

``` r
show_bits(12345L)
#> 00000000000000000011000000111001
```

# Built in annotation for ‘doubles’

``` r
plot_double(1.2345e87)
```

<img src="man/figures/README-unnamed-chunk-4-1.png" width="100%" />

``` r
plot_compact_double(1.2345e87, base_size = 5, legend.text.multiplier = 2)
```

<img src="man/figures/README-unnamed-chunk-5-1.png" width="100%" />

# Some miscellaneous examples

<img src="man/figures/README-unnamed-chunk-6-1.png" width="100%" />

<img src="man/figures/README-unnamed-chunk-7-1.png" width="100%" />

# Annotating the word ‘exhibitionist’ for the header of this page

The header of this page was made with this package.

The two key data.frames are:

  - `chars_df` which must include the columns
      - `x` for the the x-coordinate of the character
      - `char` for the character itself
      - and optional styling columns (see below
  - `annotation_df` which must include the columns
      - `start`/`end` for the start and end of a line segment underneath
        the tiles
      - `segment`/`text` logical values which indicate whether the
        segment and text should be drawn
      - and optional styling columns (see below)

Styling columns:

  - variables prefixed with `text_` are passed to `geom_text()`
  - variables prefixed with `segment_` are passed to `geom_segment()`
  - The characters inside the tiles and the annotations are styled
    independently.

<!-- end list -->

``` r
#~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
# Create the characters data.frame
#~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
chars_df <- tibble(
  char        = strsplit("Exhibitionist", '')[[1]],
  x           = seq_along(char),
  text_colour = 'black'
)

#~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
# Set the word 'bit' to be BLUE text
#~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
chars_df$text_colour[5:7] <- 'blue'

#~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
# How should the individual characters be labelled
#~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
annotation_df <- readr::read_csv(
"start, end,  segment,  text, text_size, label               , segment_colour, segment_size
     1,   4,    TRUE ,  TRUE,         7, A package for       , grey30        , 1
     5,   7,    TRUE ,  TRUE,         7, annotating          , darkgreen     , 2
     8,  13,    TRUE ,  TRUE,         7, bit & char sequences, orange        , 1
")

plot_chars(chars_df, annotation_df) + ylim(-2, 1.5)
```

<img src="man/figures/README-header-1.png" width="100%" />

# Annotating way too much

This is a demonstration of a lot more options you can change to modify
appearance.

In general:

  - variables prefixed with `text_` are passed to `geom_text()`
  - variables prefixed with `segment_` are passed to `geom_segment()`
  - The characters inside the tiles have different styling attributes
    than the annotations which appear
underneath.

<!-- end list -->

``` r
#~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
# Create the characters data.frame
#~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
chars_df <- tibble(
  char        = strsplit("Over the top", '')[[1]],
  x           = seq_along(char),
  text_colour = 'black',
  text_size   = 11,
  text_angle  = 0
)

#~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
# Style a lot of the characters
#~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
chars_df <- chars_df %>% 
  mutate(
    text_colour = if_else(x %% 2 == 0, 'red', text_colour),
    text_angle  = if_else(x %% 3 == 0,   45 , text_angle ),
    text_size   = if_else(x %% 6 == 0,   28 , text_size  ),
    tile_fill   = if_else(char == ' ', NA_character_, 'white'),
    tile_colour = if_else(char == ' ', NA_character_, 'black')
  )

#~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
# How should the individual characters be labelled underneath?
#~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
annotation_df <- readr::read_csv(
"start, end,  segment,  text, text_size, label               , segment_colour, segment_linetype, segment_size, text_family, text_angle, text_hjust
     1,   4,    TRUE ,  TRUE,        10, Over it!!           , darkgreen     , 1               , 1           , serif      , 0         ,  0.5
     6,   8,    TRUE ,  TRUE,         7, Too Much?           , blue          , 2               , 2           ,            , 45        ,  1
    10,  12,    TRUE ,  TRUE,         7, Or not enough??     , orange        , 3               , 4           ,            , 180       ,  0.5
")

plot_chars(chars_df, annotation_df) + 
  ylim(-2, 1.2)
```

<img src="man/figures/README-too_much-1.png" width="100%" />


NEWS.md
# exhibit 0.1.1

* Added `plot_compact_double()`

# exhibit 0.1.0

* Initial release


R/show-bits.R

#~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
#' Convert a single R 32-bit signed integer value to bits
#'
#' @param int32 integer value
#'
#' @return vector of 32 bits, sign bit and then 31 bits with most-significant bit first.
#~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
int32_to_bits <- function(int32) {
  stopifnot(length(int32) == 1L)
  stopifnot(is.integer(int32))
  rev(rawToBits(writeBin(int32, raw(), size = 4)))
}


#~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
#' Convert the lower 'nbits' of an int32 into bit representation
#'
#' @param int32 integer value
#' @param nbits number of lower bits to keep
#'
#' @return vector of length nbits
#~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
int_to_bits <- function(int32, nbits) {
  bits <- int32_to_bits(int32)
  rev(rev(bits)[seq(nbits)])
}


#~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
#' Convert a single double precision floating point value to bits
#'
#' @param dbl double value
#'
#' @return vector of bits, MSB first
#~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
dbl_to_bits <- function(dbl) {
  stopifnot(length(dbl) == 1L)
  stopifnot(is.double(dbl))

  rev(rawToBits(writeBin(dbl, raw(), size = 8)))
}


#~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
#' @rdname show_bits
#' @export
#~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
show_bits.double <- function(x, ...) {
  dbl <- x

  exponent_bias  <- 2^(11 - 1) - 1

  bits         <- dbl_to_bits(dbl)
  sign_bit     <- as.integer(bits[1])
  exponent     <- bits[2:12]
  exponent_str <- paste(as.integer(exponent), collapse = "")
  mantissa     <- bits[13:64]
  mantissa_str <- paste(as.integer(mantissa), collapse = "")

  exponent_val <- sum(2^(10:0) * as.integer(exponent))
  exponent_val <- exponent_val - exponent_bias


  cat(sign_bit, " ", exponent_str,
      " (", exponent_val, ") ",
      mantissa_str, " : ", dbl, "\n", sep = "")
}


#~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
#' @rdname show_bits
#' @export
#~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
show_bits.integer <- function(x, nbits = 32, ...) {
  bits <- int32_to_bits(x)
  if (nbits != 32) {
    bits <- rev(rev(bits)[seq_len(nbits)])
  }
  cat(paste(as.integer(bits), collapse = ""), "\n")
}


#~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
#' Show a representation of the bits in a numeric value
#'
#' Show a representation of the bits in a integer or floating point value
#'
#' Bits are always show with the most-significant bit on the left.
#'
#' For floating point numbers (i.e. 'doubles') the display is similar to the
#' \url{seven31}{https://github.com/ThinkR-open/seven31} package.
#'
#' @param x double or integer value
#' @param nbits number of bits to dispaly if showing an integer
#' @param ... other arguments passed to specific implementations
#'
#' @export
#~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
show_bits <- function(x, ...) {
  UseMethod('show_bits')
}

R/plot-chars.R


globalVariables(c('char', 'start', 'end', 'x', 'label',
                  'segment_y', 'segment_colour', 'segment_size', 'segment_linetype',
                  'text_y', 'text_colour', 'text_size', 'text_angle', 'text_family', 'text_hjust',
                  'tile_y', 'tile_size', 'tile_fill', 'tile_colour'))

#~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
#' Sanitise an annotation data.frame
#'
#' @param annotation_df data.frame
#' @param base_size base size
#'
#' @return sanitised annotation_df data.frame
#'
#' @import ggplot2
#~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
sanitise_annotation_df <- function(annotation_df, base_size = 11) {

  #~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
  # Sanity checks
  #~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
  if (is.null(annotation_df)) {
    return(data.frame())
  }

  stopifnot(is.data.frame(annotation_df))

  if (nrow(annotation_df) == 0L) {
    return(annotation_df)
  }

  necessary_columns <- c('start', 'end', 'segment', 'text', 'label')
  missing_columns <- setdiff(necessary_columns, colnames(annotation_df))
  if (length(missing_columns) != 0L) {
    stop("sanitise_annotation_df(): Missing essential columns: ",
         deparse(missing_columns), call. = FALSE)
  }


  #~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
  # What are sane defaults for the annotation?
  #~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
  text_defaults <- ggplot2::calc_element('text', theme_bw(base_size))
  names(text_defaults) <- paste0("text_", names(text_defaults))
  text_defaults$text_margin <- NULL

  segment_defaults <- ggplot2::calc_element('line', theme_bw(base_size))
  names(segment_defaults) <- paste0("segment_", names(segment_defaults))


  annotation_defaults <- c(
    text_defaults,
    segment_defaults
  )

  annotation_defaults$segment_y <- -0.2
  annotation_defaults$text_y    <- -0.5

  #~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
  # Apply these sane defaults when the user didn't specify a value
  # Also replace NAs with the default
  #~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
  for (ii in seq_along(annotation_defaults)) {
    col <- names(annotation_defaults)[ii]
    if (!col %in% names(annotation_df)) {
      annotation_df[[col]] <- annotation_defaults[[ii]]
    } else {
      na_idx <- is.na(annotation_df[[col]])
      annotation_df[[col]][na_idx] <- annotation_defaults[[ii]]
      class(annotation_df[[col]]) <- class(annotation_defaults[[ii]])
    }
  }

  annotation_df
}


#~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
#' Sanitise an bits data.frame
#'
#' @param chars_df data.frame
#' @param base_size base size
#'
#' @return sanitised chars_df data.frame
#'
#' @import ggplot2
#~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
sanitise_chars_df <- function(chars_df, base_size = 11) {

  #~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
  # Sanity checks
  #~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
  stopifnot(is.data.frame(chars_df))
  stopifnot(!is.null(chars_df))
  stopifnot(nrow(chars_df) > 0)

  necessary_columns <- c('char', 'x')
  missing_columns <- setdiff(necessary_columns, colnames(chars_df))
  if (length(missing_columns) != 0L) {
    stop("sanitise_chars_df(): Missing essential columns: ",
         deparse(missing_columns), call. = FALSE)
  }

  #~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
  # What are sane defaults for the MAIN TEXT + TILE ?
  #~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
  text_defaults <- ggplot2::calc_element('text', theme_bw(base_size))
  names(text_defaults) <- paste0("text_", names(text_defaults))
  text_defaults$text_margin <- NULL

  tile_defaults <- ggplot2::calc_element('rect', theme_bw(base_size))
  names(tile_defaults) <- paste0("tile_", names(tile_defaults))


  bit_defaults <- c(
    text_defaults,
    tile_defaults
  )

  bit_defaults$tile_y <- 0.5
  bit_defaults$text_y <- 0.5


  #~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
  # Apply these sane defaults when the user didn't specify a value
  # Also replace NAs with the default
  #~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
  for (ii in seq_along(bit_defaults)) {
    col <- names(bit_defaults)[ii]
    if (!col %in% names(chars_df)) {
      chars_df[[col]] <- bit_defaults[[ii]]
    } else if (!grepl('tile_colour|tile_fill', col)) {
      na_idx <- is.na(chars_df[[col]])
      chars_df[[col]][na_idx] <- bit_defaults[[ii]]
      class(chars_df[[col]]) <- class(bit_defaults[[ii]])
    }
  }

  chars_df
}




#~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
#' Plot bits
#'
#' @param chars_df data.frame with columns 'bit' and 'x'
#' @param annotation_df data.frame
#' @param title title
#' @param base_size base_size
#' @param plot_theme plot_theme
#'
#' @return ggplot object
#'
#' @import ggplot2
#' @export
#~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
plot_chars <- function(chars_df, annotation_df, title = NULL, base_size = 11,
                       plot_theme = theme_void(base_size)) {

  #~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
  # Sanitise the input data
  #~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
  chars_df      <- sanitise_chars_df(chars_df, base_size = base_size)
  annotation_df <- sanitise_annotation_df(annotation_df, base_size = base_size)


  #~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
  # Lay the Foundation of the Plot
  #~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
  p <- ggplot() +
    plot_theme +
    coord_equal()


  #~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
  # Add the tiling if requested. TRUE by default
  #~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
  p <- p +
    geom_tile(data = chars_df, aes(x      = x,
                                  y      = I(tile_y),
                                  fill   = I(tile_fill),
                                  colour = I(tile_colour),
                                  size   = I(tile_size)))


  #~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
  # Add the text for each char
  #~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
  p <- p +
    geom_text(data = chars_df, aes(x      = x,
                                   y      = I(text_y),
                                   label  = char,
                                   colour = I(text_colour),
                                   angle  = I(text_angle),
                                   size   = I(text_size)))


  #~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
  # Add a title if requested
  #~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
  if (!is.null(title)) {
    p <- p + annotate('text',
                      label = title,
                      x     = 0.5,
                      y     = 1.5,
                      hjust = 'left',
                      size  = base_size)
  }


  #~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
  # Apply the annotations - one-row-at-a-time
  #~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
  for (idx in seq(nrow(annotation_df))) {
    row <- annotation_df[idx, ]

    #~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
    # Annotate with a coloured line segment
    #~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
    if (isTRUE(row$segment)) {
      p <- p +
        geom_segment(data = row, aes(x        = start - 0.45,
                                     xend     = end   + 0.45,
                                     y        = segment_y,
                                     yend     = segment_y,
                                     colour   = I(segment_colour),
                                     linetype = I(segment_linetype),
                                     size     = I(segment_size)))
    }

    #~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
    # Annotate with text
    #~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
    if (isTRUE(row$text)) {
      p <- p + geom_text(data = row, aes(x      = 0.5 * (start + end),
                                         y      = text_y,
                                         label  = label,
                                         colour = I(text_colour),
                                         family = I(text_family),
                                         angle  = I(text_angle),
                                         hjust  = I(text_hjust),
                                         size   = I(text_size)))
    }
  }


  p
}











